File: pt5/1167.minimum-cost-to-connect-sticks.java
import java.util.*;

class Solution {
    public int connectSticks(int[] sticks) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int stick : sticks) {
            heap.offer(stick);
        }

        int total = 0;
        while (heap.size() > 1) {
            int first = heap.poll();
            int second = heap.poll();
            int merged = first + second;
            total += merged;
            heap.offer(merged);
        }

        return total;
    }
}

File: pt5/1167.minimum-cost-to-connect-sticks.go
package main

import "container/heap"

type intHeap []int

func (h intHeap) Len() int           { return len(h) }
func (h intHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h intHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *intHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *intHeap) Pop() interface{} {
	old := *h
	value := old[len(old)-1]
	*h = old[:len(old)-1]
	return value
}

func connectSticks(sticks []int) int {
	h := intHeap(sticks)
	heap.Init(&h)
	total := 0

	for h.Len() > 1 {
		first := heap.Pop(&h).(int)
		second := heap.Pop(&h).(int)
		merged := first + second
		total += merged
		heap.Push(&h, merged)
	}

	return total
}
